ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,304 @@
-   * @param size The amount of data in bytes to be written from the channel.
-   * @throws IOException if data could not be written because of I/O errors
-   *
+   * Closes the Log and all its segments.
+   * @throws IOException if the flush encountered an I/O error.
    */
-  void writeFrom(ReadableByteChannel channel, long offset, long size)
+  void close()
       throws IOException {
-    logger.trace("Log : {} currentWriteOffset {} capacityInBytes {} sizeToAppend {} offset to append at {}",
-        file.getAbsolutePath(), currentWriteOffset, capacityInBytes, size, offset);
-    if (offset < 0 || offset + size > currentWriteOffset.get()) {
-      metrics.overflowWriteError.inc(1);
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " error trying to write to log " +
-          "from channel since new data size " + size + "exceeds log end offset " + currentWriteOffset.get());
+    for (Map.Entry<String, LogSegment> entry : segmentsByName.entrySet()) {
+      entry.getValue().close();
     }
-    long bytesWritten = 0;
-    while (bytesWritten < size) {
-      bytesWritten += fileChannel.transferFrom(channel, offset + bytesWritten, size - bytesWritten);
+  }
+
+  /**
+   * Loads segment files (if they exist) and creates {@link LogSegment} instances. If no segment files exist, the first
+   * segment file is created.
+   * @param segmentCapacity the intended capacity of each log segment. This is of importance on first startup.
+   * @param numSegments the total number of segments that the log abstraction will contain.
+   * @throws IOException if there is an I/O error creating/loading the segment files or creating the {@link LogSegment}
+   * instances
+   */
+  private void loadSegments(long segmentCapacity, long numSegments)
+      throws IOException {
+    File dir = new File(dataDir);
+    File[] segmentFiles = dir.listFiles(LogSegmentNameHelper.LOG_FILE_FILTER);
+    if (segmentFiles == null || segmentFiles.length == 0) {
+      String firstSegmentName = LogSegmentNameHelper.generateFirstSegmentName(numSegments);
+      File segmentFile = allocate(LogSegmentNameHelper.nameToFilename(firstSegmentName), segmentCapacity);
+      // to be backwards compatible, headers are not written for a log segment if it is the only log segment.
+      LogSegment segment = new LogSegment(firstSegmentName, segmentFile, segmentCapacity, metrics, numSegments > 1);
+      segmentsByName.put(segment.getName(), segment);
+    } else {
+      for (File segmentFile : segmentFiles) {
+        String name = LogSegmentNameHelper.nameFromFilename(segmentFile.getName());
+        LogSegment segment;
+        if (numSegments == 1) {
+          // for backwards compatibility, a single segment log is loaded by providing capacity since the old logs have
+          // no headers
+          segment = new LogSegment(name, segmentFile, segmentCapacity, metrics, false);
+        } else {
+          segment = new LogSegment(name, segmentFile, metrics);
```

Comment:
@pnarayanan As discussed offline, there is no situation where the log segments are not the same size since the code that allocates the log segments is in the same class - so it would be a programming bug if it ended up otherwise (and there is no "fix"). Moreover, I think it is an unnecessary/misleading check that might make it seem like it could ever happen. It would be akin to checking that index and log match for example - something we do not do.